app/code/core/Mage/Adminhtml/controllers/Cms/PageController.php

```diff
@@ -102,6 +102,7 @@ public function editAction()
         // 3. Set entered data if was error when we do save
         $data = Mage::getSingleton('adminhtml/session')->getFormData(true);
         if (! empty($data)) {
+            $data['store_id'] = $data['stores'];
```

Comment:
If the form is using ```stores[]``` as the element name then it's meant to be an array. Here you seem to just be grabbing the array (or possible array) and assigning it to a field that's meant to be a scalar, no?

Reply:
Um, `store_id` i not meant to be a scalar. Think these lines from #382 make it more clear ...

    # is delete or edit action?
    $isDeleteAction = $observer->getEvent()->getName() === 'cms_page_delete_before';
    $storeIds = $isDeleteAction ? $page->getStoreId() : $page->getStores();

Reply:
I disagree.

https://github.com/OpenMage/magento-mirror/blob/magento-1.9/app/code/core/Mage/Cms/Model/Resource/Page.php#L122

The underlying model resource clearly expects that ```store_id``` is a scalar value that is used as a fallback when ```stores``` is not present. The code example you listed is faulty.

Reply:
And I'll follow up with this example that is contradictory to what I just said but which I consider a bug.

https://github.com/OpenMage/magento-mirror/blob/magento-1.9/app/code/core/Mage/Cms/Model/Resource/Page.php#L184

```store_id``` is clearly singular and in every other context on other models is a scalar. They clearly expect it to be a scalar on save since they cast it to an array. The fact that after loading they assign the array of store IDs to ```store_id``` vs. ```stores``` is, IMHO, 100% a bug in the core code.

Reply:
Further evidence that the core code is buggy regarding the treatment of ```store_id```: https://github.com/OpenMage/magento-mirror/blob/magento-1.9/app/code/core/Mage/Cms/Model/Resource/Page.php#L203-L204

Here the object loading code clearly expects ```store_id``` to be a scalar, not an array.

Reply:
It's a fallback, but for magic method `$object->getStoreId()` `store_id` can be anything.

The "problem" is that input fields name is (intentionally) `stores[]`. This form data is used when editing/saving CMS pages. When delete a page `getData()` contains the DB fields, so it is `store_id`.

> The code example you listed is faulty. 
Why?

However, fix works :)

Reply:
> IMHO, 100% a bug in the core code.

Thats what i've meant. :)

Reply:
The input name is ```stores``` because that is the data field that is expected when saving a page. The fact that the core afterLoad code is dumping the store IDs into ```store_id``` instead of ```stores``` is clearly a bug. There is no DB field for ```store_id``` of a page (directly). There is an additional link table that links pages to stores so a page has an array of store ids always.

Reply:
Isn't it posible to add an alias to `cms_page_store.store_id as stores`?

(But then you need a fallback to `store_id`, at least for 3rd party code.)

Reply:
I would say adding ```stores``` in the after load is a good first step in the overall issue.

Reply:
Adding `$object->setData('stores', $stores);` to `_afterLoad` solves only my problem (different getters for store ids), but it doesnt solve the problem @colinmollenhour describes.

To reproduce ...

- add an observer to `cms_page_save_before`-
- just throw an exeption to display an error message

Selected stores are still empty.